amaranth/back/rtlil.py

```diff
@@ -47,221 +48,349 @@ def _const(value):
+        self.attributes = {"generator": "Amaranth"}
         self.emit_src = emit_src
+        if src_loc is not None and emit_src:
+            self.attributes["src"] = _src(src_loc)
+        if attrs is not None:
+            self.attributes.update(attrs)
 
-    def module(self, name=None, attrs={}, *, src=None):
-        name = self._make_name(name, local=False)
-        return _ModuleBuilder(self, name, attrs, src=src)
+    def _anon_name(self):
+        self._anon_index += 1
+        return f"${self._anon_index}"
 
+    def wire(self, width, *, name=None, **kwargs):
+        if name is None:
+            name = self._anon_name()
+        else:
+            name = f"\\{name}"
+        assert name not in self.contents
+        if not self.emit_src and "src_loc" in kwargs:
+            del kwargs["src_loc"]
+        self.contents[name] = res = Wire(width, name=name, **kwargs)
+        return res
 
-class _ModuleBuilder(_AttrBuilder, _BufferedBuilder, _Namer):
-    def __init__(self, rtlil, name, attrs, *, src=None):
-        super().__init__(emit_src=rtlil.emit_src)
-        self.rtlil = rtlil
-        self.name  = name
-        self.src   = src
-        self.attrs = {"generator": "Amaranth"}
-        self.attrs.update(attrs)
+    def cell(self, kind, name=None, **kwargs):
+        if name is None:
+            name = self._anon_name()
+        else:
+            name = f"\\{name}"
+        assert name not in self.contents
+        if not self.emit_src and "src_loc" in kwargs:
+            del kwargs["src_loc"]
+        self.contents[name] = res = Cell(kind, name=name, **kwargs)
+        return res
 
-    def __enter__(self):
-        self._attributes(self.attrs, src=self.src)
-        self._append("module {}\n", self.name)
-        return self
+    def memory(self, width, depth, name=None, **kwargs):
+        if name is None:
+            name = self._anon_name()
+        else:
+            name = f"\\{name}"
+        assert name not in self.contents
+        if not self.emit_src and "src_loc" in kwargs:
+            del kwargs["src_loc"]
+        self.contents[name] = res = Memory(width, depth, name=name, **kwargs)
+        return res
 
-    def __exit__(self, *args):
-        self._append("end\n")
-        self.rtlil._buffer.write(str(self))
+    def process(self, *, name=None, **kwargs):
+        if name is None:
+            name = self._anon_name()
+        else:
+            name = f"\\{name}"
+        assert name not in self.contents
+        if not self.emit_src and "src_loc" in kwargs:
+            del kwargs["src_loc"]
+        self.contents[name] = res = Process(name=name, **kwargs)
+        return res
 
-    def wire(self, width, port_id=None, port_kind=None, name=None, attrs={}, src="", signed=False):
+    def connect(self, lhs, rhs):
+        self.connections.append((lhs, rhs))
+
+    def attr(self, name, value):
+        assert name not in self.attributes
+        self.attributes[name] = value
+
+    def emit(self, emit):
```

Comment:
`emit(emit)` reads really weird.

```suggestion
    def emit(self, write):
```